coremasscheck.F90

```diff
@@ -0,0 +1,66 @@
+! Include shortname defintions, so that the F77 code does not have to be modified to
+! reference the CARMA structure.
+#include "carma_globaer.h"
+
+!!  This routine checks if the coremass exceeds the total
+!!
+!! @author Yunqian Zhu
+!! @version Apr-2021
+subroutine coremasscheck(carma, cstate, iz, fixcoremass,logmsg,abort, packagename, rc)
+
+  ! types
+  use carma_precision_mod
+  use carma_enums_mod
+  use carma_constants_mod
+  use carma_types_mod
+  use carmastate_mod
+  use carma_mod
+
+  implicit none
+
+  type(carma_type), intent(inout)      :: carma       !! the carma object
+  type(carmastate_type), intent(inout) :: cstate      !! the carma state object
+  integer, intent(in)                  :: iz
+  logical, intent(in)                  :: fixcoremass !! if we fix the coremass if exceeds total
+  logical, intent(in)                  :: logmsg      !! write message to log file if core mass check fails
+  logical, intent(in)                  :: abort       !! set return code to ERROR and return if  if core mass check fails
+  character(len=*), intent(in)         :: packagename !! string to add to error message
+  integer, intent(inout)               :: rc          !! return code, negative
+
+  ! Local declarations
+  integer                        :: igroup,ibin
+  integer                        :: iepart,i,icore
+  real(kind=f)                   :: coremass
+  real(kind=f)                   :: factor
+
+  real(kind=f), parameter :: thresh  = 1e-10_f     ! relative threshold for core mass and kappa checking
+  real(kind=f), parameter :: thresh1 = 1._f+thresh
+
+  ! check the coremass exceeding the total mass
+  do igroup = 1,NGROUP
+     iepart = ienconc(igroup)
+     do ibin = 1,NBIN
+        if (pc(iz, ibin, iepart) > 0._f) then
+           coremass = 0._f
+           do i = 1, ncore(igroup)
+              icore = icorelem(i,igroup)
+              coremass = coremass + pc(iz, ibin, icore)
+           end do ! i = 1, ncore(igroup)
+           if (coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
```

Comment:
This seems like it will allow cases where coremass is greater than pc by a little, which could generate negative values for the concentration element. Seems like you should apply a threshold the other way if at all to make sure you never get negative values for the concentration element.

Reply:
@zywshoon

Reply:
@cbardeen  This is because if core mass is greater than the total by a little bit (i.e. caused by the round-up ), we don't need to fix it. and the round up I set it to 1e-10.

Reply:
But I think you do need to fix it, since other people that don't check and just do pc * rmass - sum(coremass) will find a negative number that will probably not do the right thing in their calculation if it bigger by up to 1e-10. I also think that 1e-10 is larger than a round off error.

Reply:
@cbardeen Are you suggesting that we reduce the number 1e-10 to a smaller value?
Or are you suggesting to fix always when coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) ?

Reply:
is done (code change by Yunqian)

Reply:
Don't think this is right. Should always fix if core mass > PC, but can fix PC to a value that is threshold larger than core mass to keep it from failing again later.